src/main/java/edu/harvard/iq/dataverse/engine/command/impl/UpdateDatasetVersionCommand.java

```diff
@@ -165,38 +167,77 @@ public Dataset execute(CommandContext ctxt) throws CommandException {
+            theDataset = ctxt.em().merge(theDataset);
 
+            /*
+             * This code has to handle many cases, and anyone making changes should
+             * carefully check tests and basic methods that update the dataset version. The
+             * differences between the cases stem primarily from differences in whether the
+             * files to add, and there filemetadata, and files to delete, and their
+             * filemetadata have been persisted at this point, which manifests itself as to
+             * whether they have id numbers or not, and apparently, whether or not they
+             * exists in lists, e.g. the getFileMetadatas() list of a datafile.
+             *
+             * To handle this, the code is carefully checking to make sure that deletions
+             * are deleting the right things and not, for example, doing a remove(fmd) when
+             * the fmd.getId() is null, which just removes the first element found.
+             */
             for (FileMetadata fmd : filesToDelete) {
+                logger.fine("Deleting fmd: " + fmd.getId() + " for file: " + fmd.getDataFile().getId());
+                // if file is draft (ie. new to this version), delete it. Otherwise just remove
+                // filemetadata object)
+                // There are a few cases to handle:
+                // * the fmd has an id (has been persisted) and is the one in the current
+                // (draft) version
+                // * the fmd has an id (has been persisted) but it is from a published version
+                // so we need the corresponding one from the draft version (i.e. created during
+                // a getEditVersion call)
+                // * the fmd has no id (hasn't been persisted) so we have to use non-id based
+                // means to identify it and remove it from lists
+
+                if (fmd.getId() != null) {
+                    // If the datasetversion doesn't match, we have the fmd from a published version
+                    // and we need to remove the one for the newly created draft instead, so we find
+                    // it here
+                    if (theDataset.getEditVersion() != fmd.getDatasetVersion()) {
+                        fmd = FileMetadataUtil.getFmdForFileInEditVersion(fmd, theDataset.getEditVersion());
+                    }
+                } else {
+                    // Not sure if this is needed now that there is a dataset merge above, but we
+                    // need to assure it is on the context
+                    fmd = ctxt.em().merge(fmd);
+                }
+                // There are two datafile cases as well - the file has been released, so we're
+                // jsut removing it from the current draft version or it is only in the draft
+                // version and we completely remove the file.
                 if (!fmd.getDataFile().isReleased()) {
```

Comment:
Getting a null pointer here when I try to delete a file via the EditDataFilesPage from the Dataset Page. Will investigate further.

Reply:
@sekmiller - The new commits should fix this - the test of the two dataset versions in line 208 needed to use ! .equals() instead of !=, and then an fmd merge was needed (removing the else around line 212.

Reply:
Ok. i'm able to delete a file from a draft version. Thanks!